microsoft/pxt
# Avoid Cross Site Scripting Vectors To/From the Simulator Iframe

### Problem
The simulator is loaded in an iframe and a separate domain from the editor/page that hosts it. We communicate with the simulator via the _postMessage()_ command, and receive messages via a message event listener. When we post messages to the simulator iframe, we currently pass the _*_ wild card as the _targetOrigin_. When we receive messages from the simulator iframe, we don't check that the origin matches the domain of the simulator iframe. 

### Solution
The SimulatorDriver is run in the parent window of the simulator iframe. It facilitates communication between the simulator iframe, the page that loaded the SimulatorDriver, and even it's parent window. The latter case comes into play when you visit a page like https://makecode.com/multi, which can load multiple editor pages with simulators that will be able to communicate with one another. The solution comes in a few parts:
- Include the origin of the parent window as a URL search parameter when loading in editor in either makecode.com/multi or makecode.com/streamer
- Check that the origin of messages sent to the editor/SimulatorDriver match either the _simUrl_ of the simulator iframe, the editor's origin, or the _parentOrigin_ if that is specified.
- Use the URL of the simulator, _simUrl_, when _postMessage_ is called on the simulator iframe
- Use the origin of the editor, or the _parentOrigin_, when broadcasting message up to the parent window
- Use the origin of the editor when broadcasting messages to dependent editors

### Tests
I did lots of manual testing to make sure that messages were still going back and forth as expected. 
- Check that broadcasting still works by using multi (I had to manually edit the multi.html to use my changes to include the parent origin in the editor's URI) to load two Arcade pages; one page includes a radio broadcaster and the other a radio listener.
- Did the same on the test as above on the streamer page with no issues.
- Check that broadcasting between an editor and a dependent editor works by opening a dependent editor in a new tab that received messages from the current tab (the inverse does not work even before my changes).
- Check that simulator to simulator communication still works by creating a single project with a radio broadcaster and radio listener. This will create two simulators, both changing their color based on the broadcasted values.
- Check that editor to simulator communication still works via taking a screenshot/recording the simulator in the share modal.
- I couldn't test my changes with the share page, but that doesn't listen to messages like the streamer or multi pages do. So, the editor running in the iframe should run normally. Once this change is in, I can update one of the targets, say maker.makecode.com, to use the new pxt version to confirm the share project page works as expected.

### Note
This is only half of the problem of cross scripting vectors. The next step would be to put the origin of the editor page as a search parameter in the URI of the simulator iframe. The simulator iframe would then check against that editor origin when it receives or sends messages to the editor. Should I create a separate bug to track this, or should I consider the bug unresolved?

Here is the link to uploaded MakeCode Arcade that I used to test my changes: https://arcade.makecode.com/app/39fe5a831cd7b837be2e61d6f2f65e9ce59a347c-f5305f6e21

Fixes https://github.com/microsoft/pxt-backend/issues/109
